### ISSUE TITLE
runtime question

### ISSUE THREAD

**Issue by mkinnaman** [ChristofferFlensburg/superFreq#49]
Hi!

I had a quick question about setting runtime on my initial run of the pipeline. I am trying to run on two normals plus 6 tumor samples. From my reading of the tutorial - it should take 24 hours for whole genome for a cancer-normal set. So should I estimate 24 hours per sample Running on 24 cores)? Does memory request scale with the number of samples as well or should 15 GB a core (360gb total) be sufficient? 

Finally anyway to pick up where you left off if job killed? It keeps restarting with the first sample when examining x positions in y sample.

Thanks,
Michael

Thanks,
Michael

**Comment by ChristofferFlensburg** [comment]
Hi Michael!

Just to be clear, you have 8 matched samples, 6 cancer and two normals all from the same individual? Seems like an awesome dataset! :D

24h is for a single human cancer with a matched normal on 10 cores and 200GB memory. More samples will take more time, as well as requiring more memory. Essentially the memory scales with the number of positions studied (from the input VCF), the number of samples, and roughly the number of cpus. So if you run this on 10 cpus, it should finish in a few days.

So if you want to limit memory use:
- decrease the number of cpus. Try 10?
- try to be a little more conservative in the preliminary variant calling. We usually get around 5M variants in the input VCF for genomes, 1-2M for inbred lab mice. If you are much above that, there should be improvements to make.

So first thing I'd try is to make sure that the number of positions (your "x") isn't exceptionally large, and decrease to 10 cpus, keep 360GBs. Or go 20 cpus and increase memory if you have the resources and are in a hurry. Let me know how it goes!

Memory also varies from case to case, depending on noise levels and number of germline variants and so on.

There are save points (that are loaded by default), but none between samples, only after it completed the variant calling of all samples. That might be a good thing to add in future version though! Trade-off is disk space, but should be small compared to the BAM.

In general, I am aware superFreq starts to use a lot of resources for many matched human genomes. Part of that is just R not coping well with heavy calculations, but I am working on improving it, and getting genomes to scale better is high on priority list. Unfortunately I am also stretched between many different projects that are all ranging between "high" and "urgent" priority. 😬

**Comment by mkinnaman** [comment]
Hi Chris,

Sorry for the delayed response - I am actually still running this first case - but we are slowly getting to the finish line. Excited to see the results.

I am also trying to run this on a pair of targeted sequencing bam files. I used your varscan script to call variants and then tried to run using the captureRegions argument to specify the targets. 

I keep running into this same error message:

`Error in names(assigned) <- sampleNames : 
  'names' attribute [3] must be the same length as the vector [2]
Calls: superFreq -> superFreq -> analyse -> runDE
In addition: Warning messages:
1: In read.table(sampleMetaDataFile, header = T, as.is = T, fill = T,  :
  incomplete final line found by readTableHeader on '/home/kinnamam/superfreq/superfreq_input6/rms_target.tsv'
2: In runDE(bamFiles, sampleNames, externalNormalCoverageBams, captureRegions,  :
  Saved normal counts dont match the bam files curently in the normal directory. Please rerun with "forceRedo$forceRedoNormalCount = T" to redo the normal counts on the current bam files.
Execution halted`

Not sure if this can even be run on targeted sequencing (468 gene panel) but thought I would give it a shot.

Let me know what you think - I cant figure out how to add "forceRedo$forceRedoNormalCount = T"  to the Rscript but that just probably highlights my novice R skills.

Thanks,
Michael

**Comment by ChristofferFlensburg** [comment]
Hi!

The very first draft of superFreq was actually made for a capture set of 630 genes, and while the algorithm certainly has changed since, it should still handle it well. Since it's pretty rare, it not well supported in terms of easy-of-use though. :/ Essentially you have two options:

1) The easiest way is to just run them as exomes, only that most genes will have close to 0 coverage. Variant calling should handle this fine, but the copy number calling might get confused.
2) Using custom capture regions, which should produce better copy number calls, but might require some manual work to get running.

These two produce different read count matrices of genes x samples (for the copy number calling), as 1) will look at ~20k genes, 2) will look at 468. I think what happened is that your capture region run is fed a bed file with 468 genes, but the reference normal directory already has a saved count matrix with more genes.

If you haven't used R a lot, then the easiest way to redo the counts is probably to just delete the save file `normalFCsExon.Rdata` file in the R subdirectory of your normal directory.

Note that you can't use the same normal directory with different capture regions for this reason. So if you run with the 1) approach you need to set up a different normal directory from the 2) approach. You can link the bam files though, no need to copy.

I hope I answered your question, let me know how it goes!

stay safe
/Christoffer

**Comment by mkinnaman** [comment]
Thanks Chritoffer,

I was able to figure out how to put forceredo into the Rscript so I tried running again using option 2 as you described above, deleting the normalFCsExon.Rdata file and I got held up by this error:

`Error in if (floor(ranks[i]) < currentRank) next : 
  missing value where TRUE/FALSE needed
Calls: superFreq -> superFreq -> analyse -> runDE -> XRank -> postRank`

Thoughts ?

**Comment by ChristofferFlensburg** [comment]
Hi!

I can't immediately guess what is causing that error message, but it seems related to the read counts over the capture regions.

I would first try option 1 and see if the results are ok.

Otherwise, I would need to look at the logfile (Rdirectory/[individual]/runtimetracking.log), and perhaps also the bed files and the count matrix (Rdirectory/[individual]/fCsExon.Rdata). If you are up for it we could also try zoom (or equivalent) and maybe we can get your analysis running quicker.

**Comment by mkinnaman** [comment]
I was able to bypass that error running as exome but running into an issue with no reads being found at variant locations. I think its a problem on my end - running the targeted data through a different pipeline and will let you know how it goes.

One other question I had - if I had a patient with both targeted panel sequencing and genome sequencing from different time points - could I, in theory, run this on exome mode on those samples?

**Comment by ChristofferFlensburg** [comment]
OK, good. Let me know if you can't get it to find the reads.

For mixing data types: it is possible to give the BAM files as input in theory, and if you run in exome mode it'll just ignore reads from the genome that are not close to an exon. The problem is which reference normal you give it, as it would not match one of the samples. I think the variant calls might work decently, but the copy number calls will be very weird. Probably better to use genome reference normals if you have them, to get better coverage of the variants. 🤔 I haven't done it myself (I run separately and then merge/compare later), so I honestly don't know if it'll give you anything sensible. If you try, I'd be curious to hear how it goes!

**Comment by mkinnaman** [comment]
Christoffer -

I cant see to get past this error using targeted sequencing data:
`Error in FUN(X[[i]], ...) : 
  Found no reads over any positions in the VCF. This is typically a sign of failing to read from the bam (/juno/work/isabl/public/kinnaman/kinnaman_20200221/KD641286-T.bam), bam index or fasta (/juno/work/isabl/ref/homo_sapiens/GRCh37d5/genome/gr37.fasta). Or it can be an issue with samtools not being properly available.`

I have tried running it multiple different ways but cant get past this error.  I don't think its an issue with the bams/reference either. Not sure if you have any other ideas.

Thanks,
Michael

**Comment by ChristofferFlensburg** [comment]
Hi!

I just recently found out that superFreq isn't compatible with the latest version of samtools 1.10 (due to different column order in mpileup). So might be that. If so, run 1.9 or earlier, or wait until I fixed it, hopefully next week.

It might also be a case of a first run having issues and saving corrupted data that is then being reused (look for and remove anything with "variants" in it in the Rdirectory/[individual], then rerun).

If neither of those two, then send me the logfile (Rdirectory/[individual]/runtimeTracking.log), and I'll have a closer look for clues!

Thanks for patience!

**Comment by mkinnaman** [comment]
Hey Christoffer,

Appreciate your responsiveness - I was hoping it would be as simple as the samtools version problem but I have been using 1.7 and have had no issues running genomes so far so that hasn't been the issue. I always wipe all directories clean before each run so I don't think its that either. Attached is the runtimelog - let me know what you think.

[runtimeTracking.log](https://github.com/ChristofferFlensburg/superFreq/files/4488308/runtimeTracking.log)

**Comment by ChristofferFlensburg** [comment]
Hi!

The logfile does find samtools 1.7, and it seems to read from the bam files correctly for the counts, so prime suspects are the bam index files, the fasta file, or some bug inside superFreq parsing. If you used the same fasta for the genomes, then I guess that is ruled out as well.

If you think that those files are in order, then next step is running again (you can run in same directory if you want), setting `cpus=1`. The error is inside a multi-threaded part, which doesn't pass error messages properly in R, but running on a single thread seems to often give more meaningful messages.

**Comment by ChristofferFlensburg** [comment]
I've had to ask a lot of users to do this, I think I should automate it, and tell superFreq to do this as automated follow up instead of throwing the error message you got...

**Comment by mkinnaman** [comment]
here is a log file from cpu=1, not sure if there is any more insight from this...
[runtimeTracking.log](https://github.com/ChristofferFlensburg/superFreq/files/4490647/runtimeTracking.log)

**Comment by ChristofferFlensburg** [comment]
Thanks!

Unfortunately the error message is the same, so no new information. :/

Especially because you have run it before, I suspected the error is related to the input files. Let's try to reproduce it on the command line (outside of R), by doing the same samtools mpiluep call superFreq does. So outside of R (with samtools loaded as usual), run:

```
samtools mpileup -OsB -f [pathToFasta] -r [chr:pos:pos] [bam]
```

where the fasta file is the file you give to superFreq (which should be the fasta you aligned to), and [chr:pos:pos] are taken from any variant in the input VCF. SuperFreq internally strips all "chr" from the chromosome names, but for the samtools call it re-adds it if needed, so use the chromosome names you find in the VCF. Hopefully this should give you a more informative error message, or the output might have clues to why the parsing fails. Just paste the output (and error message) back here.

Thanks, hopefully this will solve it!

**Comment by mkinnaman** [comment]
Christoff,

Happy to report this issue has been solved. To make a long story short I didn't notice the vcfs were made off recalibrated BAM files that were stored in a different directory. Once the correct recalibrated bams were identified the pipeline ran to completion! Really came out quite nice I think. Thanks again for being so responsive.

Kind Regards,
Michael

**Comment by ChristofferFlensburg** [comment]
Ok, that's great! 🎉 

This was the panel data, right? I'm curious if the copy numbers came out ok despite being run as exomes, if you don't mind sharing. :)

**Comment by mkinnaman** [comment]
Yes - this was our 468 gene targeted panel (MSK IMPACT). I took a quick look at the clinical report and the copy number calls/profiles seemed to match up pretty well. Our reports don't call copy neutral LOH so in some areas where that was called by superfreq they were just reported as losses on the clinical report. WIll let you know if anything comes up as I dig deeper but on the surface they look pretty good.

**Comment by ChristofferFlensburg** [comment]
Ok, sounds good, thanks!

Can I close this issue? You can always create a new ticket if needed.

**Comment by mkinnaman** [comment]
Yes thanks!
